README.md
# waxjs

Javascript API for integration with the WAX Cloud Wallet.

## Documentation

[Check the WAX developer portal for full instructions](https://developer.wax.io/waa/waxjs-overview/)

## Installation

### Browser

Grab the waxjs.js minified bundle in the dist-web folder of this repo, or build it yourself (see below). [Check the example code](https://github.com/worldwide-asset-exchange/waxjs/blob/develop/dist-web/index.html) to see how to use it.

### NPM

```bash
npm install @waxio/waxjs
```

### YARN

```bash
yarn add @waxio/waxjs
```

## Usage

### 0. Import (for NPM and Yarn installations)

React style apps using npm or yarn can import the library via:

```js
import * as waxjs from "@waxio/waxjs/dist";
```

### 1. Instantiate

Instantiate the waxjs object with the RPC server you wish to connect to.

```js
const wax = new waxjs.WaxJS('https://wax.greymass.com');
```

The library can also be instatiated with the user account and the public keys. Due to the library contains the user information, the login step can be avoided. 

```js
const wax = new waxjs.WaxJS('https://wax.greymass.com', 'user1', ['EOS7rC8jFvFrPYDqp3Nh3HdRfL79h11B1JhPEXy85enF5wwYzF3Hk']);
```

If you want to handle the auto-login on your side with the ```isAutoLoginAvailable``` function (to avoid waiting for the user to click a button), you can disable the auto-login function in the constructor (so it won't get called twice).

```js
const wax = new waxjs.WaxJS('https://wax.greymass.com', null, null, false);
```
 
### 2. Login

Log your user in so as to access their wax account name for creating transactions.

```js
const userAccount = await wax.login();
```

Successful login will return the userAccount. It will also be available as the ```userAccount``` member on the ```wax``` instance. You can now use the eosjs ```api``` member...


### 3. Use the eosjs Objects As Usual

Utilize the eosjs ```api``` and ```rpc``` members available on the ```wax``` instance. They are instances of the regular eosjs objects, Api, and JsonRpc, so you can do anything with them that eosjs already provides. [Check the eosjs docs and repo](https://github.com/EOSIO/eosjs) for more info.

The api method will not be initialized until you login your user, and remember that the user's account name is available as the ```userAccount``` member on the ```wax``` instance.

```js
const result = await wax.api.transact({
  actions: [{
    account: 'eosio.token',
    name: 'transfer',
    authorization: [{
      actor: wax.userAccount,
      permission: 'active',
    }],
    data: {
      from: wax.userAccount,
      to: 'eosio',
      quantity: '0.00000001 WAX',
      memo: '',
    },
  }]
}, {
  blocksBehind: 3,
  expireSeconds: 1200,
});
```

## Development

### Generate docs

``` npm run docs```

### Run tests

``` npm run test```

### Build lib

``` npm run build```

### Build for web 

``` npm run build-web```

## Contributing

When making a pull request, please make sure to run `npm run prettier` to make sure your code is as formatted as possible. Also, make sure `npm run lint` runs without errors, since that is the final check before a new version is published to npm.


CHANGELOG.md
# Changelog

## Pending

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS:

BUG FIXES:

## v0.0.15

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS:

BUG FIXES:
- Fix for "new transaction" and "not iterable" errors when signing transactions

## v0.0.14

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS:

BUG FIXES:
- Improved error reporting when signing fails or is dismissed by the user

## v0.0.13

BREAKING CHANGES:

FEATURES:
- Optional signing capacity for an API provider to pay for a user's CPU/NET via ONLY_BILL_FIRST_AUTHORIZER

IMPROVEMENTS:

## v0.0.12

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS:
- don't fail on alternate origin

## v0.0.11

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS: 
- [KEW-1816] Attempt to auto-login when waxjs is instantiated

BUG FIXES:

## v0.0.10

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS:

BUG FIXES:
- [KEW-1806] Fix default params for typescript and make api member public

## v0.0.9

BREAKING CHANGES:
- [KEW-1804] Accept initialization with account and pubkeys

FEATURES:

IMPROVEMENTS:

BUG FIXES:

## v0.0.8

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS:
- [KEW-1801] Only autologin on chrome

BUG FIXES:
- Fix for whitelists not getting set by some browsers. Ex Brave

## v0.0.7

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS:
- [KEW-1776] Whitelisted actions and auto sign-in

BUG FIXES:

## v0.0.6

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS:

BUG FIXES:
- lint for publishing

## v0.0.5

BREAKING CHANGES:

FEATURES:

IMPROVEMENTS:

BUG FIXES:
- [KEW-1724] Opening sign dialog not working on safari

## v0.0.4

BREAKING CHANGES:

FEATURES:
- [KEW-1582] Project Boilerplate
- [KEW-1706] defaults point to prod instances, reorder init params, cloud-wallet path, api tweaks
- [KEW-1701] docs

IMPROVEMENTS:
- [KEW-1716] npom and yarn installation

BUG FIXES:
- [] Fix multiple signatures
- [KEW-1726] Login not working on firefox


test/Example.test.ts
import { expect } from 'chai';

describe('My Example', () => {
  it('should return hello wax', () => {
    expect('Hello wax').to.equal('Hello wax');
  });
});

src/WaxEventSource.ts
export class WaxEventSource {
  constructor(private waxSigningURL: string = "http://localhost:3000") {
    this.openEventSource = this.openEventSource.bind(this);
    this.onceEvent = this.onceEvent.bind(this);
  }

  public async openEventSource(
    url: string,
    message?: any,
    win?: Window
  ): Promise<any> {
    const openedWindow = win
      ? win
      : await window.open(url, "WaxPopup", "height=800,width=600");

    if (!openedWindow) {
      throw new Error("Unable to open a popup window");
    }

    if (typeof message === "undefined") {
      return openedWindow;
    }

    const postTransaction = async (event: any) => {
      if (event.data.type === "READY") {
        // @ts-ignore
        openedWindow.postMessage(message, this.waxSigningURL);
      }
    };

    const eventPromise = this.onceEvent(
      // @ts-ignore
      openedWindow,
      this.waxSigningURL,
      postTransaction
    );

    await Promise.race([eventPromise, this.timeout()]).catch(err => {
      if (err.message !== "Timeout") {
        throw err;
      }

      openedWindow.postMessage(message, this.waxSigningURL);
    });

    return openedWindow;
  }

  public async onceEvent(
    source: Window,
    origin: string,
    action: (e: any) => void
  ) {
    return new Promise((resolve, reject) => {
      (window as Window).addEventListener(
        "message",
        async function onEvent(event) {
          // Validate expected origin for event
          if (event.origin !== origin) {
            return;
          }

          // Validate expected source for event
          if (event.source !== source) {
            return;
          }

          if (typeof event.data !== "object") {
            return;
          }

          try {
            const result: any = await action(event);
            resolve(result);
          } catch (e) {
            reject(e);
          }

          (window as Window).removeEventListener("message", onEvent, false);
        },
        false
      );
    });
  }

  private timeout = () => {
    return new Promise((resolve, reject) => {
      const wait = setTimeout(() => {
        clearTimeout(wait);
        reject(new Error("Timeout"));
      }, 2000);
    });
  };
}


src/index.ts
import { Api, JsonRpc } from "eosjs";
import { SignatureProvider } from "eosjs/dist/eosjs-api-interfaces";
import { IWhitelistedContract } from "./IWhitelistedContract";
import { WaxEventSource } from "./WaxEventSource";
import * as waxjs from "@waxio/waxjs/dist";


export class WaxJS {
  public api: Api;
  private waxEventSource: WaxEventSource;
  private readonly rpc: JsonRpc;
  private userAccount: string;
  private pubKeys: string[];
  private signingWindow: Window;
  private whitelistedContracts: IWhitelistedContract[];

  constructor(
    rcpEndpoint: string,
    userAccount: string = null,
    pubKeys: string[] = null,
    tryAutoLogin: boolean = true,
    private apiSigner: SignatureProvider = null,
    private waxSigningURL: string = "https://all-access.wax.io",
    private waxAutoSigningURL: string = "https://api-idm.wax.io/v1/accounts/auto-accept/"
  ) {
    this.waxEventSource = new WaxEventSource(waxSigningURL);
    this.rpc = new JsonRpc(rcpEndpoint);

    if (userAccount && Array.isArray(pubKeys)) {
      // login from constructor
      const data = { userAccount, pubKeys, verified: true };
      this.receiveLogin({ data });
    } else {
      // try to auto-login via endpoint
      if (tryAutoLogin) {
        this.loginViaEndpoint();
      }
    }
  }

  public async login() {
    if (this.userAccount && Array.isArray(this.pubKeys)) {
      return this.userAccount;
    } else {
      // login via UI
      return this.loginViaWindow();
    }
  }

  public async isAutoLoginAvailable() {
    if (this.userAccount && Array.isArray(this.pubKeys)) {
      return true;
    } else {
      // try to auto-login via endpoint
      try {
        await this.loginViaEndpoint();
        return true;
      } catch (e) {
        return false;
      }
    }
    return false;
  }

  private async loginViaWindow() {
    const confirmationWindow = await this.waxEventSource.openEventSource(
      this.waxSigningURL + "/cloud-wallet/login/"
    );

    return this.waxEventSource.onceEvent(
      confirmationWindow,
      this.waxSigningURL,
      this.receiveLogin.bind(this)
    );
  }

  private async loginViaEndpoint() {
    const response = await fetch(this.waxAutoSigningURL + "login", {
      credentials: "include",
      method: "get"
    });
    if (!response.ok) {
      throw new Error(
        `Login Endpoint Error ${response.status} ${response.statusText}`
      );
    }
    const data = await response.json();
    if (data.processed && data.processed.except) {
      throw new Error(data);
    }
    return this.receiveLogin({ data });
  }

  private async receiveLogin(event: any) {
    const {
      verified,
      userAccount,
      pubKeys,
      whitelistedContracts,
      autoLogin
    } = event.data;
    if (!verified) {
      throw new Error("User declined to share their user account");
    }

    if (userAccount == null || pubKeys == null) {
      throw new Error("User does not have a blockchain account");
    }

    localStorage.setItem("autoLogin", autoLogin);
    this.whitelistedContracts = whitelistedContracts || [];
    this.userAccount = userAccount;
    this.pubKeys = pubKeys;

    const signer = {
      getAvailableKeys: async () => {
        return [
          ...this.pubKeys,
          ...((this.apiSigner && (await this.apiSigner.getAvailableKeys())) ||
            [])
        ];
      },
      sign: async (data: any) => {
        return {
          serializedTransaction: data.serializedTransaction,
          signatures: [
            ...(await this.signing(data.serializedTransaction)),
            ...((this.apiSigner &&
              (await this.apiSigner.sign(data)).signatures) ||
              [])
          ]
        };
      }
    };
    // @ts-ignore
    this.api = new Api({ rpc: this.rpc, signatureProvider: signer });
    const transact = this.api.transact.bind(this.api);
    const url = this.waxSigningURL + "/cloud-wallet/signing/";
    // We monkeypatch the transact method to overcome timeouts
    // firing the pop-up which some browsers enforce, such as Safari.
    // By pre-creating the pop-up window we will interact with,
    // we ensure that it is not going to be rejected due to a delayed
    // pop up that would otherwise occur post transaction creation
    this.api.transact = async (transaction, namedParams) => {
      if (!(await this.canAutoSign(transaction))) {
        this.signingWindow = await window.open(
          url,
          "WaxPopup",
          "height=800,width=600"
        );
      }
      return await transact(transaction, namedParams);
    };

    return this.userAccount;
  }

  private async canAutoSign(transaction: any) {
    const deserializedTransaction = transaction.actions
      ? transaction
      : await this.api.deserializeTransactionWithActions(transaction);
    return !deserializedTransaction.actions.find((action: any) => {
      return !this.isWhitelisted(action);
    });
  }

  private isWhitelisted(action: any) {
    return !!this.whitelistedContracts.find((w: any) => {
      if (w.contract === action.account) {
        if (action.account === "eosio.token" && action.name === "transfer") {
          return w.recipients.includes(action.data.to);
        }
        return true;
      }
      return false;
    });
  }

  private async signing(transaction: any) {
    if (await this.canAutoSign(transaction)) {
      return this.signViaEndpoint(transaction).catch(() =>
        // Attempt to recover by signing via the window method
        this.signViaWindow(undefined, transaction)
      );
    }

    return this.signViaWindow(this.signingWindow, transaction);
  }

  private async signViaEndpoint(transaction: any) {
    try {
      const response: any = await fetch(this.waxAutoSigningURL + "signing", {
        body: JSON.stringify({
          transaction: Object.values(transaction)
        }),
        credentials: "include",
        headers: {
          "Content-Type": "application/json"
        },
        method: "POST"
      });
      if (!response.ok) {
        throw new Error(
          `Signing Endpoint Error ${response.status} ${response.statusText}`
        );
      }
      const data: any = await response.json();
      if (data.processed && data.processed.except) {
        throw new Error(data);
      }
      return this.receiveSignatures({ data });
    } catch (e) {
      // clear the whitelist to make sure we don't repeatedly attempt blocked actions
      this.whitelistedContracts = [];
      throw e;
    }
  }

  private async signViaWindow(window: Window, transaction: any) {
    const confirmationWindow: Window = await this.waxEventSource.openEventSource(
      this.waxSigningURL + "/cloud-wallet/signing/",
      { type: "TRANSACTION", transaction },
      window
    );

    return this.waxEventSource.onceEvent(
      confirmationWindow,
      this.waxSigningURL,
      this.receiveSignatures.bind(this)
    );
  }

  private async receiveSignatures(event: any) {
    if (event.data.type === "TX_SIGNED") {
      const { verified, signatures, whitelistedContracts } = event.data;
      if (!verified || signatures == null) {
        throw new Error("User declined to sign the transaction");
      }
      this.whitelistedContracts = whitelistedContracts || [];

      return signatures;
    } else if (event.data.type !== "READY") {
      throw new Error(
        `Unexpected response received when attempting signing: ${JSON.stringify(
          event.data,
          undefined,
          2
        )}`
      );
    }
    return [];
  }
}
